build-tools/packages/build-cli/src/commands/generate/sourceEntrypoints.ts

```diff
@@ -0,0 +1,386 @@
+import {
+	readPackageJson,
+	readTsConfig,
+	// AB#8118 tracks removing the barrel files and importing directly from the submodules, including disabling this rule.
+	// eslint-disable-next-line import/no-internal-modules
+} from "../../library/package.js";
+// AB#8118 tracks removing the barrel files and importing directly from the submodules, including disabling this rule.
+// eslint-disable-next-line import/no-internal-modules
+import type { ExportData } from "../../library/packageExports.js";
+// AB#8118 tracks removing the barrel files and importing directly from the submodules, including disabling this rule.
+// eslint-disable-next-line import/no-internal-modules
+import { getApiExports, getPackageDocumentationText } from "../../library/typescriptApi.js";
+import type { CommandLogger } from "../../logging.js";
+
+/**
+ * Generates source entrypoints for Fluid Framework APIs to support API levels (/alpha, /beta. etc.).
+ */
+export default class GenerateSourceEntrypointsCommand extends BaseCommand<
+	typeof GenerateSourceEntrypointsCommand
+> {
+	static readonly description =
+		`Generates TypeScript source files that roll up APIs into different entrypoint files, defined by the "exports" field in package.json and organized by API tags`;
+
+	static readonly flags = {
+		mainEntrypoint: Flags.file({
+			description: "Main entrypoint file containing all untrimmed exports.",
+			default: "./src/index.ts",
+			exists: true,
+		}),
+		outDir: Flags.directory({
+			description: "Directory to emit entrypoint files.",
+			default: "./src/entrypoints/",
+			exists: true,
+		}),
+		...BaseCommand.flags,
+	};
+
+	public async run(): Promise<void> {
+		const { mainEntrypoint, outDir } = this.flags;
+
+		const packageJson = await readPackageJson();
+
+		// Read tsconfig present at the root of the package
+		const tsConfig = await readTsConfig();
+
+		const mapApiTagToExportPath: Map<ApiTag, ExportData> = mapExportPathToApiTag(
+			packageJson,
+			tsConfig,
+			this.logger,
+		);
+
+		const mapApiTagToSourcePath: Map<string, ApiTag> = await mapSourcePathToApiTag(outDir);
+
+		const mapSourceToExportPath: Map<ApiTag, ExportData> = getOutputConfiguration(
+			mapApiTagToSourcePath,
+			mapApiTagToExportPath,
+			this.logger,
+		);
+
+		if (mapSourceToExportPath.size === 0) {
+			throw new Error(
+				`There are no package exports matching requested output entrypoints:\n\t${[
+					...mapApiTagToSourcePath.keys(),
+				].join("\n\t")}`,
+			);
+		}
+
+		// generate source entrypoints under `${outDir}`
+		return generateSourceEntrypoints(mainEntrypoint, mapSourceToExportPath, this.logger);
+	}
+}
+
+// Formats the given output directory path to ensure it starts with `./` and ends with `/`.
+function formatPath(outDir: string): string {
+	return `./${outDir.replace(/^\.\/|\/$/g, "")}/`;
+}
+
+/**
+ * Read and parse tsconfig under `outDir`
+ * @returns rootDir
+ */
+async function readOutDirTsConfig(outDir: string): Promise<string> {
```

Comment:
name is a bit confusing - at first I thought it was going to read the `outDir` from the tsconfig. Instead it reads the `rootDir` from a tsconfig in a path given in argument named `outDir`.

General naming philosophy advice - name a thing for what is does and not how you intend to use it.

In the end I expect that this will need to report both `rootDir` and `outDir`. Perhaps call if `readTsConfigDirectories`. The argument called `outDir` can be renamed to `tsconfigDir` (assuming it shouldn't just accept a full path, which I'd name `tsconfigPath` (recommended) ).

Reply:
The command is reading two tsconfigs:
1. One at the root of the package 
2. tsconfig present under `outDir` or `./src/entrypoints`

I think it's not required to read both since `emitDeclarationOnly` will be set to true in both tsconfigs.  If that's okay, looking at option 2, then the two pieces of info we need from tsconfig is `rootDir` and `emitDeclarationOnly`. Renaming this function to `getTsConfigRootDirAndEmitOptions`